crates/air-lib/air-parser/src/parser/validator.rs

```diff
@@ -30,6 +30,194 @@ use multimap::MultiMap;
+                    }
+                    _ => {
+                        // disable machine if Xoring invariant, namely there must be 2 kinds on a stack, is broken.
+                        self.stack.clear();
+                        self.is_enabled = false;
+                    }
+                }
+            }
+            CheckInstructionKind::Merging => {
+                let right_branch = self.stack.pop();
+                let left_branch = self.stack.pop();
+                match (left_branch, right_branch) {
+                    (
+                        Some((CheckInstructionKind::PivotalNext(left_iterable), _)),
+                        Some((CheckInstructionKind::PivotalNext(right_iterable), _)),
+                    ) if left_iterable == right_iterable => {
+                        self.stack
+                            .push((CheckInstructionKind::PivotalNext(left_iterable), span));
+                    }
+                    (Some((CheckInstructionKind::PivotalNext(iterator), _)), Some((..))) => {
+                        // potential failure but need to check when fold pops up.
+                        self.stack.push((CheckInstructionKind::Merging, span));
+                        self.potentially_impossible_spans.push((iterator, span));
+                    }
+                    (Some((..)), Some((CheckInstructionKind::PivotalNext(iterable), _))) => {
+                        self.stack
+                            .push((CheckInstructionKind::PivotalNext(iterable), span));
+                    }
+                    (Some(_), Some(_)) => {
+                        self.stack.push((CheckInstructionKind::Merging, span));
+                    }
+                    _ => {
+                        // disable machine if Merging invariant, namely there must be 2 kinds on a stack, is broken.
+                        self.stack.clear();
+                        self.is_enabled = false;
+                    }
+                }
+            }
+            CheckInstructionKind::ReplacingWithCheck(iterator_name) => {
+                self.replacing_with_check_common(iterator_name, span, &instr_kind);
+            }
+            CheckInstructionKind::PopStack1ReplacingWithCheck(iterator_name) => {
+                self.stack.pop();
+                self.replacing_with_check_common(iterator_name, span, &instr_kind);
+            }
+            CheckInstructionKind::PivotalNext(_) | CheckInstructionKind::Simple => {
+                self.stack.push((instr_kind, span))
+            }
+            CheckInstructionKind::PopStack1 => {
+                self.stack.pop();
+                self.stack.push((instr_kind, span))
+            }
+            CheckInstructionKind::PopStack2 => {
+                self.stack.pop();
+                self.stack.pop();
+                self.stack.push((instr_kind, span))
+            }
+        }
+    }
+
+    fn after_next_check(&mut self, iterable: &'name str) {
```

Comment:
Has same name as a field, makes it less convenient read.

Reply:
Renamed